cmd/xl-v1-multipart-common.go

```diff
@@ -20,169 +20,147 @@ import (
+				// default (empty) upload info.
+				uploadsJSON, err = newUploadsV1("xl"), nil
 			}
-			if wErr = disk.AppendFile(minioMetaBucket, tmpUploadsPath, uploadsBytes); wErr != nil {
-				errs[index] = traceError(wErr)
+			// If we have a read error, we store error and
+			// exit.
+			if err != nil {
+				errs[index] = traceError(err)
 				return
 			}
-			if wErr = disk.RenameFile(minioMetaBucket, tmpUploadsPath, minioMetaBucket, uploadsPath); wErr != nil {
-				errs[index] = traceError(wErr)
-				return
+
+			if !uCh.isRemove {
+				// Add the uploadID
+				uploadsJSON.AddUploadID(
+					uCh.uploadID,
+					uCh.initiated,
+				)
+			} else {
+				// Remove the upload ID
+				uploadsJSON.RemoveUploadID(uCh.uploadID)
+				if len(uploadsJSON.Uploads) == 0 {
+					isDelete[index] = true
+				}
+			}
+
+			// Check if there are entries to write.
+			if !isDelete[index] {
+				// Serialize to prepare to write to disk.
+				uploadsJSONBytes, wErr := json.Marshal(&uploadsJSON)
+				if wErr != nil {
+					errs[index] = traceError(wErr)
+					return
+				}
+
+				// Write `uploads.json` to disk. First to tmp
+				// location and then rename.
+				if wErr = disk.AppendFile(minioMetaBucket, tmpUploadsPath, uploadsJSONBytes); wErr != nil {
+					errs[index] = traceError(wErr)
+					return
+				}
+				wErr = disk.RenameFile(minioMetaBucket, tmpUploadsPath, minioMetaBucket, uploadsPath)
+				if wErr != nil {
+					if dErr := disk.DeleteFile(minioMetaBucket, tmpUploadsPath); dErr != nil {
+						// we store the most recent error.
+						errs[index] = traceError(dErr)
+						return
+					}
+					errs[index] = traceError(wErr)
+				}
+			} else {
+				// If there are no entries we prepare
+				// to delete uploads.json from
+				// disk. We do this by first renaming
+				// the file to tmp location and then
+				// purging if we reach quorum
+				wErr := disk.RenameFile(minioMetaBucket, uploadsPath, minioMetaBucket, tmpUploadsPath)
+				if wErr != nil {
+					errs[index] = traceError(wErr)
+				}
+
 			}
 		}(index, disk)
 	}
 
-	// Wait for all the routines to finish updating `uploads.json`
+	// Wait for all the writes to finish.
 	wg.Wait()
 
-	// Count all the errors and validate if we have write quorum.
+	// Do we have write quorum?.
 	if !isDiskQuorum(errs, xl.writeQuorum) {
-		// Do we have readQuorum?.
-		if isDiskQuorum(errs, xl.readQuorum) {
-			return nil
-		}
-		// Rename `uploads.json` left over back to tmp location.
-		for index, disk := range xl.storageDisks {
-			if disk == nil {
+		// No quorum. Perform cleanup on the minority of disks
+		// in which the operation succeeded.
+		for index, disk := range diskPermutation {
+			if disk == nil || errs[index] != nil {
 				continue
 			}
 			// Undo rename `uploads.json` in parallel.
 			wg.Add(1)
 			go func(index int, disk StorageAPI) {
 				defer wg.Done()
-				if errs[index] != nil {
-					return
+
+				if !isDelete[index] {
+					// we don't want a partially
```

Comment:
Don't manually CR comments as i said before there is no 80 line requirement in Golang. This kind of looks ugly. This comment can be made better simply by saying 

```
// Remove partially written files.
```